cmd/lockrelease/main.go

```diff
@@ -0,0 +1,122 @@
+)
+
+const (
+	gceInstanceIDKey = "container.googleapis.com/instance_id"
+	leaseName        = "filestore-csi-storage-gke-io-node"
+	// Root CA configmap in each namespace.
+	rootCA = "kube-root-ca.crt"
+)
+
+var (
+	lockReleaseSyncPeriod = flag.Duration("lock-release-sync-period", 60*time.Second, "Duration, in seconds, the sync period of the lock release controller. Defaults to 60 seconds.")
+
+	httpEndpoint = flag.String("http-endpoint", "", "The TCP network address where the HTTP server for diagnostics, including metrics and leader election health check, will listen (example: `:8080`). The default is empty string.")
+	metricsPath  = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+
+	leaderElectionLeaseDuration = flag.Duration("leader-election-lease-duration", 15*time.Second, "Duration, in seconds, that non-leader candidates will wait to force acquire leadership. Defaults to 15 seconds.")
+	leaderElectionRenewDeadline = flag.Duration("leader-election-renew-deadline", 10*time.Second, "Duration, in seconds, that the acting leader will retry refreshing leadership before giving up. Defaults to 10 seconds.")
+	leaderElectionRetryPeriod   = flag.Duration("leader-election-retry-period", 5*time.Second, "Duration, in seconds, the LeaderElector clients should wait between tries of actions. Defaults to 5 seconds.")
+)
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatalf("Failed to create an in cluster config: %v", err)
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		klog.Fatalf("Failed to create a new discovery client: %v", err)
+	}
+	lockReleaseConfig := &releaselock.LockReleaseControllerConfig{
+		LeaseDuration:  *leaderElectionLeaseDuration,
+		RenewDeadline:  *leaderElectionRenewDeadline,
+		RetryPeriod:    *leaderElectionRetryPeriod,
+		SyncPeriod:     *lockReleaseSyncPeriod,
+		MetricEndpoint: *httpEndpoint,
+		MetricPath:     *metricsPath,
+	}
+	factory := informers.NewSharedInformerFactory(client, lockReleaseConfig.SyncPeriod)
+	nodeInformer := factory.Core().V1().Nodes().Informer()
+
+	c, err := releaselock.NewLockReleaseController(client, lockReleaseConfig, &nodeInformer)
+	if err != nil {
+		klog.Fatalf("Failed to create a lock release controller: %v", err)
+	}
+
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			klog.Infof("Node informer received node create event. %v", obj)
+			c.EnqueueCreateEventObject(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			klog.Infof("Node informer received node update event. old %v, new %v", oldObj, newObj)
+			c.EnqueueUpdateEventObject(oldObj, newObj)
+		},
+	})
+
+	run := func(ctx context.Context) {
+		klog.Infof("Lock release controller %s started leading on node %s", c.GetId(), c.GetHost())
```

Comment:
I saw the same log in `releaselock/controller.go`. Are both required?

Reply:
ack